kernel: work: Fix race in work queue thread

Songjf-ttk opened the pull request:
After a call to `k_work_flush` returns the sync variable may still be modified by the workq.  This is because the work queue thread continues to modify the flag in sync even after `k_work_flush` returns.  This commit adds
`K_WORK_FLUSHING_BIT`, and with this bit, we moved the logic of waking up the caller from `handle_flush` to the `finalize_flush_locked` in workq, so that after waking up the caller, the workqueue will no longer operate on sync.

Fixes [#64530](https://github.com/zephyrproject-rtos/zephyr/issues/64530)

Signed-off-by: Junfan Song <sjf221100@gmail.com>

andyross commented on kernel/work.c:
```diff
@@ -671,19 +671,31 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
+		 * sync for synchronization will return. At this point,
+		 * both the calling thread and the work queue thread can
+		 * access sync. To ensure the consistency of sync, it is not
+		 * allowed for the work queue thread to modify sync anymore.
```
This seems incomplete?  This is going to skip the call to finalize_cancel_locked(), which as I read it is doing important stuff like list management and wakeups.  And I don't see where that handling is being done in other contexts.  Why is this safe?

andyross reviewed (commented):
Some questions after a first read.  The work queue state management is indeed sorta... ponderously complicated.  And I totally buy that the linked bug is real.  I do worry that this fix isn't complete.  It looks like it fixes the modification race but leaves other stuff dangling?

Songjf-ttk committed:
kernel: work: Fix race in workqueue thread

After a call to k_work_flush returns the sync variable
may still be modified by the workq.  This is because
the work queue thread continues to modify the flag in
sync even after k_work_flush returns.  This commit adds
K_WORK_FLUSHING_BIT, and with this bit, we moved the
logic of waking up the caller from handle_flush to the
finalize_flush_locked in workq, so that after waking up
the caller, the workqueue will no longer operate on sync.

Fixes: #64530

Signed-off-by: Junfan Song <sjf221100@gmail.com>

Songjf-ttk commented:
Hi @andyross, thanks for your review.  I have now modified the code, changing `K_WORK_SYNC_BIT` to `K_WORK_FLUSHING_BIT`,  specifically to fix the bugs introduced by flush.  To eliminate the inconsistency issues with sync,  I have delayed the call to `k_sem_give` from `handle_flush` to `finalize_flush_locked`.  This way,  after `k_work_flush` returns,  the workqueue thread will no longer modify sync.  Additionally,  I have added a check for `K_WORK_FLUSHING_BIT` in the workqueue,  so as not to affect the logic of the existing code:
```
if (flag_test(&work->flags, K_WORK_FLUSHING_BIT)) {
	finalize_flush_locked(work);
}
```

andyross reviewed (dismissed):
This makes sense to me, thanks (with the caveat that races are hard and the fact that I can't see anything wrong doesn't mean it's not!).  FWIW: I salute your work in generating that flowchart, though think maybe it's a little overkill for the bug. :)

andyross reviewed (approved):
Refresh +1
